File: body tracking/multi-camera/cpp/src/ClientPublisher.cpp
#include "ClientPublisher.hpp"

ClientPublisher::ClientPublisher() { }

ClientPublisher::~ClientPublisher()
{
    zed.close();
}

bool ClientPublisher::open(sl::InputType input, Trigger* ref) {

    p_trigger = ref;

    sl::InitParameters init_parameters;
    init_parameters.depth_mode = sl::DEPTH_MODE::ULTRA;
    init_parameters.input = input;
    init_parameters.coordinate_units = sl::UNIT::METER;
    init_parameters.depth_stabilization = 5;
    auto state = zed.open(init_parameters);
    if (state != sl::ERROR_CODE::SUCCESS)
    {
        std::cout << "Error: " << state << std::endl;
        return false;
    }

    serial = zed.getCameraInformation().serial_number;
    p_trigger->states[serial] = false;

    // in most cases in body tracking setup, the cameras are static
    sl::PositionalTrackingParameters positional_tracking_parameters;
    // in most cases for body detection application the camera is static:
    positional_tracking_parameters.set_as_static = true;
    
    state = zed.enablePositionalTracking(positional_tracking_parameters);
    if (state != sl::ERROR_CODE::SUCCESS)
    {
        std::cout << "Error: " << state << std::endl;
        return false;
    }

    // define the body tracking parameters, as the fusion can does the tracking and fitting you don't need to enable them here, unless you need it for your app
    sl::BodyTrackingParameters body_tracking_parameters;
    body_tracking_parameters.detection_model = sl::BODY_TRACKING_MODEL::HUMAN_BODY_MEDIUM;
    body_tracking_parameters.body_format = sl::BODY_FORMAT::BODY_18;
    body_tracking_parameters.enable_body_fitting = false;
    body_tracking_parameters.enable_tracking = false;
    state = zed.enableBodyTracking(body_tracking_parameters);
    if (state != sl::ERROR_CODE::SUCCESS)
    {
        std::cout << "Error: " << state << std::endl;
        return false;
    }

    return true;
}

void ClientPublisher::start()
{
    if (zed.isOpened()) {
        // the camera should stream its data so the fusion can subscibe to it to gather the detected body and others metadata needed for the process.
        zed.startPublishing();
        // the thread can start to process the camera grab in background
        runner = std::thread(&ClientPublisher::work, this);
    }
}

void ClientPublisher::stop()
{
    if (runner.joinable())
        runner.join();
    zed.close();
}

void ClientPublisher::work()
{    
    sl::Bodies bodies;
    sl::BodyTrackingRuntimeParameters body_runtime_parameters;
    body_runtime_parameters.detection_confidence_threshold = 40;

    sl::RuntimeParameters rt;
    rt.confidence_threshold = 50;

    // in this sample we use a dummy thread to process the ZED data.
    // you can replace it by your own application and use the ZED like you use to, retrieve its images, depth, sensors data and so on.
    // as long as you call the grab function and the retrieveBodies (which runs the detection) the camera will be able to seamlessly transmit the data to the fusion module.
    while (p_trigger->running) {
        std::unique_lock<std::mutex> lk(mtx);
        p_trigger->cv.wait(lk);
        if(p_trigger->running){
         if(zed.grab(rt) == sl::ERROR_CODE::SUCCESS){
            // just be sure to run the bodies detection
            zed.retrieveBodies(bodies, body_runtime_parameters);
         }
        }
        p_trigger->states[serial] = true;
    }
}

void ClientPublisher::setStartSVOPosition(unsigned pos) {
    zed.setSVOPosition(pos);
}

File: body tracking/multi-camera/csharp/ClientPublisher.cs
using sl;
using System;
using System.Threading;
using static Khronos.Platform;

class ClientPublisher
{
    sl.Camera zedCamera;
    Thread thread;
    bool running = false;
    int id = 0;

    /// <summary>
    /// 
    /// </summary>
    public ClientPublisher(int id_)
    {
        id = id_;
        running = false;
        zedCamera = new Camera(id_);
    }

    /// <summary>
    /// 
    /// </summary>
    /// <param name="inputType"></param>
    /// <returns></returns>
    public bool Open(sl.InputType inputType)
    {
        if (thread != null && thread.IsAlive)  return false;

        sl.InitParameters initParameters = new sl.InitParameters();
        initParameters.resolution = sl.RESOLUTION.AUTO;
        initParameters.cameraFPS = 30;
        initParameters.depthMode = sl.DEPTH_MODE.ULTRA;
        initParameters.coordinateUnits = sl.UNIT.METER;
        initParameters.coordinateSystem = COORDINATE_SYSTEM.RIGHT_HANDED_Y_UP;
        initParameters.sdkVerbose = 1;
        
        switch(inputType.GetType())
        {
            case sl.INPUT_TYPE.USB:
                initParameters.inputType = sl.INPUT_TYPE.USB;
                initParameters.cameraDeviceID = id;
                break;
            case sl.INPUT_TYPE.SVO:
                initParameters.inputType = sl.INPUT_TYPE.SVO;
                initParameters.pathSVO = new string(inputType.svoInputFilename);
                break;
            case sl.INPUT_TYPE.STREAM:
                initParameters.inputType = sl.INPUT_TYPE.STREAM;
                initParameters.ipStream = new string(inputType.streamInputIp);
                initParameters.portStream = inputType.streamInputPort;
                break;
            case sl.INPUT_TYPE.GMSL:
                initParameters.inputType = sl.INPUT_TYPE.GMSL;
                break;
            default:
                Console.WriteLine("ERROR: Invalid input type");
                return false;
        }

        ERROR_CODE err = zedCamera.Open(ref initParameters);

        if (err != ERROR_CODE.SUCCESS)
        {
            Console.WriteLine("ERROR while opening the camera. Exiting...");
            Environment.Exit(-1);
        }


        if (zedCamera.CameraModel == MODEL.ZED)
        {
            Console.WriteLine(" ERROR : not compatible camera model");
            Environment.Exit(-1);
        }

        // Enable tracking (mandatory for body trackin)
        PositionalTrackingParameters positionalTrackingParameters = new PositionalTrackingParameters();
        err = zedCamera.EnablePositionalTracking(ref positionalTrackingParameters);

        if (err != ERROR_CODE.SUCCESS)
        {
            Console.WriteLine("ERROR while enable the positional tracking module. Exiting...");
            Environment.Exit(-1);
        }

        // Enable the Objects detection module
        sl.BodyTrackingParameters bodyTrackingParameters = new BodyTrackingParameters();
        bodyTrackingParameters.enableObjectTracking = false; // the body tracking will track bodies across multiple images, instead of an image-by-image basis
        bodyTrackingParameters.enableSegmentation = false;
        bodyTrackingParameters.enableBodyFitting = false; // smooth skeletons moves
        bodyTrackingParameters.detectionModel = sl.BODY_TRACKING_MODEL.HUMAN_BODY_MEDIUM;
        bodyTrackingParameters.bodyFormat = sl.BODY_FORMAT.BODY_38;
        bodyTrackingParameters.allowReducedPrecisionInference = true;
        err = zedCamera.EnableBodyTracking(ref bodyTrackingParameters);

        if (err != ERROR_CODE.SUCCESS)
        {
            Console.WriteLine("ERROR while enable the body tracking module. Exiting...");
            Environment.Exit(-1);
        }

        return true;
    }

    /// <summary>
    /// 
    /// </summary>
    public void Start()
    {
        if (zedCamera.IsOpened())
        {
            running = true;
            sl.CommunicationParameters communicationParameters = new sl.CommunicationParameters();
            communicationParameters.communicationType = sl.COMM_TYPE.INTRA_PROCESS;
            ERROR_CODE err =zedCamera.StartPublishing(ref communicationParameters);
            if (err != ERROR_CODE.SUCCESS)
            {
                Console.WriteLine("ERROR while startPublishing" + err + " . Exiting...");
                Environment.Exit(-1);
            }
            thread = new Thread(new ThreadStart(Work));
            thread.Priority = System.Threading.ThreadPriority.Highest;
            thread.Start();
        }
    }

    /// <summary>
    /// 
    /// </summary>
    public void Stop()
    {
        running = false;
        if (thread != null && thread.IsAlive)
        {
            thread.Join();
        }
        zedCamera.Close();
    }

    /// <summary>
    /// 
    /// </summary>
    /// <param name="pos"></param>
    public void SetStartSVOPosition(int pos)
    {
        zedCamera.SetSVOPosition(pos);
    }

    /// <summary>
    /// 
    /// </summary>
    /// <returns></returns>
    public bool IsRunning()
    {
        return running;
    }

    /// <summary>
    /// 
    /// </summary>
    private void Work()
    {
        sl.Bodies bodies = new sl.Bodies();
        sl.BodyTrackingRuntimeParameters bodyTrackingRuntimeParameters = new sl.BodyTrackingRuntimeParameters();
        bodyTrackingRuntimeParameters.detectionConfidenceThreshold = 40;

        sl.RuntimeParameters runtimeParameters = new sl.RuntimeParameters();
        sl.ERROR_CODE err = ERROR_CODE.FAILURE;
        while (IsRunning())
        {
            err = zedCamera.Grab(ref runtimeParameters);
            if (err == sl.ERROR_CODE.SUCCESS)
            {
                err = zedCamera.RetrieveBodies(ref bodies, ref bodyTrackingRuntimeParameters);
            }
            else
            {
                Console.WriteLine("Error while grabbing: " + err);
            }
            Thread.Sleep(2);
        }
    }
}